sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/StreamExecution.scala

```diff
@@ -71,9 +71,18 @@ class StreamExecution(
   /** The current batchId or -1 if execution has not yet been initialized. */
   private var currentBatchId: Long = -1
 
+  private[sql] val logicalPlan = _logicalPlan.transform {
+    case StreamingRelation(sourceCreator, output) =>
+      // Materialize source to avoid creating it in every batch
+      val source = sourceCreator()
+      // We still need to use the previous `output` instead of `source.schema` as attributes in
+      // "_logicalPlan" has already used attributes of the previous `output`.
+      StreamingRelation(() => source, output)
```

Comment:
I tried `Map[DataSource, Source]` but failed because of RichSource.

```
  implicit class RichSource(s: Source) {
    def toDF(): DataFrame = Dataset.ofRows(sqlContext, StreamingRelation(s))

    def toDS[A: Encoder](): Dataset[A] = Dataset(sqlContext, StreamingRelation(s))
  }
```
If we only have `StreamingRelaction(DataSource)`, then RichSource needs to create a DataSource for Source dynamically. 

So the above codes will be changed to
```
  implicit class RichSource(s: Source) {
    def toDF(): DataFrame = Dataset.ofRows(sqlContext, StreamingRelation(DataSource(sqlContext, className = ...)))

    def toDS[A: Encoder](): Dataset[A] = Dataset(sqlContext, StreamingRelation(sqlContext, className = ...))
  }
```

Here I don't what to fill for `className`. Without code generation, we won't be able to create a new class for different Source instances. This seems too complicated.

Therefore, I used the `StreamExecutionRelation` idea finally.